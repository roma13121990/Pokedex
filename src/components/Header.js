import React, {Component} from 'react';
import {Link} from "react-router-dom";

const styles = {

    'background': '#ffffff',
    'padding': 16,
    'fontSize': '1.5em',
    'color': '#515673',
    'textAlign': 'left',
    'position': 'fixed',
    'left':0,
    'right':0,
    'top': 0,
    'zIndex': 99

};
const button = {
    "float": "right",
    "textDecoration": "none",
    "fontSize": "1rem",
    "color": "#ffffff",
    "background": "#f66",
    "padding": "5px 10px",
    "borderRadius": "19px"
}
export default class Header extends Component {
    render() {
        return (<div style={styles}>Pokedex
                <Link style={button} to='/MyPokemonList'>My Pokemon</Link>
            </div>

        )
    }
}