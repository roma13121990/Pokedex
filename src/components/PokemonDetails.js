import React,{ Component } from 'react'
import '../styles/styles.css'
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import WebApi from "../utils/WebApi";
import Loader from 'react-loader'

export class PokemonDetails extends Component{
    constructor(props){
        super(props);

        this.state= {
            loading: false,
            data: {
                stats: [],
                moves: []
            }
        }
    }
    componentWillReceiveProps(nextProps){
        this.setState({
            data: nextProps.data
        })
    }

   componentDidMount() {
       WebApi.getEachPokemon(this.props.match.params.name).then(res => res.json())
           .then(response => {
               this.setState({
                   loading: true,
                   data: response
               })
           });
    }


    render() {
        const { data,loading } = this.state;
        return (<Loader loaded={loading}>
            <div className="tabs">
                <Tabs>
                    <TabList>
                        <Tab>Info</Tab>
                        <Tab>Stats</Tab>
                        <Tab>Moves</Tab>
                    </TabList>

                    <TabPanel>
                        <ul className="tab-content">
                            <li><label>Name</label><span>{data.name}</span></li>
                            <li><label>Height</label><span>{data.height}</span></li>
                            <li><label>Weight</label><span>{data.weight}</span></li>
                        </ul>

                    </TabPanel>
                    <TabPanel>
                        <ul className="tab-content">{data.stats.map((a, i) =>
                            <li key={i}><label>{a.stat.name}</label><span>{a.base_stat}</span></li>
                        )}

                        </ul>
                    </TabPanel>
                    <TabPanel>
                        <ul className="tab-content">{data.moves.map((a, i) =>
                            <li key={i}><label>{a.move.name}</label></li>
                        )}

                        </ul>
                    </TabPanel>

                </Tabs>
            </div></Loader>


        )

    }
}