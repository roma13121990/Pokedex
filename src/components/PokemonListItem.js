import React, {Component} from 'react'
import {addPokemon, removePokemon} from "../actions/action";
import {connect} from "react-redux";
import {toast} from "react-toastify";
import 'react-toastify/dist/ReactToastify.css';
import FaPlus from 'react-icons/lib/fa/plus'

//Same component is used for showing all pokemon and My pokemon list
export class PokemonListItem extends Component {

    constructor(props) {
        super(props);
        this.state = {
            type: []
        }
    }

    //takes name in small case as input and return first letter capital
    capitalizeFirstLetter = (name) => {
        return name.slice(0, 1).toUpperCase() + name.slice(1, name.length);
    };

    //dispatches add action if pokemon does not exist in MyPokemon
    addToList = (pokemon) => {
        let text = this.capitalizeFirstLetter(pokemon.name)
        if (this.checkLocalStoreForItem()) {
            //dispatch add action
            this.props.dispatch(addPokemon(pokemon))
            //show toast if pokemon is successfully added
            toast(text + " added successfully !");
        }
        else {
            //show toast if pokemon is already in the list
            toast(text + " is already yours !");
        }
    };

    removeFromList = (name) => {
        //dispatch delete event
        this.props.dispatch(removePokemon(name))
    };

    checkLocalStoreForItem = () => {
        //Checks localstorage to see if pokemon is already added
        const {pokemon} = this.props;
        const serializedState = localStorage.getItem('state');
        if(serializedState) {
            const getItem = JSON.parse(serializedState).item;
            if (getItem.some(item => item.pokemon.name
                    === pokemon.name)) {
                return false
            }
            else {
                return true
            }
        }else {
            return true
        }

    };

    render() {
        const {pokemon, showBtn} = this.props;
        //Add button only displayed if pokemon is not in the list
        //showBtn is passed as props from My Pokemon/ all pokemon to determine which button to display

        let hideAddBtn = this.checkLocalStoreForItem();

        let text = this.capitalizeFirstLetter(pokemon.name)
        return <div className="pokemon--species">
            <ul className="pokemon--species--container">
                <li className="pokemon--species--name">{text}</li>
                <li className="pokemon--species--sprite">
                    <div><img src={require(`../../public/pokemons/${pokemon.name}.jpg`)} title={pokemon.name}
                              alt={pokemon.name}/>
                        {(hideAddBtn && showBtn === "add") && <button className="action" onClick={(e) => {
                            e.preventDefault();
                            this.addToList(pokemon)
                        }}><FaPlus/></button>}
                        {showBtn === "remove" && <button className="action" onClick={(e) => {
                            e.preventDefault();
                            this.removeFromList(pokemon.name)
                        }}>X</button>}
                    </div>
                </li>
            </ul>
        </div>;
    }
}

export default connect()(PokemonListItem)