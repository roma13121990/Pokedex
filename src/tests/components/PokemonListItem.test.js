import React from 'react';
import { shallow, mount } from "enzyme";
import Adapter from 'enzyme-adapter-react-16';
import * as Enzyme from "enzyme";
import {PokemonListItem} from "../../components/PokemonListItem";
import renderer from "react-test-renderer";
import {MemoryRouter} from "react-router-dom";

const props = {'pokemon':{'name':"pikachu"}}

Enzyme.configure({ adapter: new Adapter() });
describe('<PokemonListItem />', () => {
    beforeEach(() => {
        //mock localStorage

        global.localStorage = {
            getItem: function () {
                return JSON.stringify({'item' : [
                        {'pokemon':{'name':"Pikachoo"}},
                        {'pokemon':{'name':"Abc"}},
                        {'pokemon':{'name':"xyz"}}
                    ]})
            }
        };
    });

    it("snapshot from pokemon list with data from localstore", () => {
        const tree = renderer
            .create(<MemoryRouter>
                <PokemonListItem {...props}/>
            </MemoryRouter>)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should return capitalized first letter of a name', () => {

        const wrapper = shallow(<PokemonListItem {...props}/>)
        expect(wrapper.instance().capitalizeFirstLetter("pokemon")).toEqual('Pokemon');

    })

})