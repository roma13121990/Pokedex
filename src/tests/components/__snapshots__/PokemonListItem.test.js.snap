// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonListItem /> snapshot from pokemon list with data from localstore 1`] = `
<div
  className="pokemon--species"
>
  <ul
    className="pokemon--species--container"
  >
    <li
      className="pokemon--species--name"
    >
      Pikachu
    </li>
    <li
      className="pokemon--species--sprite"
    >
      <div>
        <img
          alt="pikachu"
          src="pikachu.jpg"
          title="pikachu"
        />
      </div>
    </li>
  </ul>
</div>
`;
